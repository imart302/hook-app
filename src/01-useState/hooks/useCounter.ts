import { useState } from "react"


export const useCounter = (initial: number) => {

  const [count, setCount] = useState<number>(initial);


  const increment = () => {
    setCount((current) => current + 1);
  }

  const decrement = () => {
    setCount((current) => current - 1);
  }

  const reset = () => {
    setCount(initial);
  }

  return {count, increment, decrement, reset};

} 