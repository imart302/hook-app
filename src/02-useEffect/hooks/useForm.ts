import React, { ChangeEvent, useState } from 'react'

export interface IUseFormSte {
  username: string,
  email: string,
  password: string,
}

export default function useForm(initialForm: IUseFormSte = {username: '', email:'', password: ''}) {
  
  const [state, setState] = useState<IUseFormSte>(initialForm);


  const inputChange = (event: ChangeEvent<HTMLInputElement>) => {
    const {name, value} = event.target;

    setState({
      ...state,
      [name]: value
    });

  };

  const resetForm = () => {
    setState(initialForm);
  }
  
  return (
    {
      form: state,
      inputChange,
      resetForm,
    }
  )
}
