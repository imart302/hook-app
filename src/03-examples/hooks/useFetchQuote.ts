import React, { isValidElement, useEffect, useState } from 'react';
import { getQuote } from '../../api/breaking_bad/getQuotes';

export interface IUseFetchQuoteSte {
  data: any,
  isLoading: boolean,
  error: Error | null | string,
  fetchController: AbortController | null,
}

export interface IUseFetchQuoteProps {
  index: number
}

export const useFetchQuote = (props: IUseFetchQuoteProps) => {


  const [state, setState] = useState<IUseFetchQuoteSte>({
    data: {},
    isLoading: true,
    error: null,
    fetchController: null,
  });

  const getQuoteA = async () => {
    try{

      //abort previus fetch if exists
      if(state.fetchController && state.isLoading){
        state.fetchController.abort();
        setState({
          ...state,
          isLoading: true,
          fetchController: null,
        });
      }

      const abortController = new AbortController();
      setState({
        ...state,
        isLoading: true,
        fetchController: abortController,
      });

      const resp = await getQuote(props.index, abortController);
      const data = await resp.json();
      if(data.length === 0) {
        throw new Error('No Data');
      }
      console.log(data); 

      setState({
        data,
        isLoading: false,
        error: null,
        fetchController: null,
      });
    }
    catch(error){
      setState({
        data: {},
        isLoading: false,
        error: (error instanceof Error) ? error : 'error',
        fetchController: null
      });
    }
  }

  useEffect(() => {
    getQuoteA();
  }, [props.index]);

  return {
    data: state.data,
    isLoading: state.isLoading,
    error: state.error,
  };
};
