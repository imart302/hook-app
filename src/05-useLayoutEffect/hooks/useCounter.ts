import React, { useState } from 'react';

export interface IUseCounterSte {
  count: number;
}

export interface IUseCounterProps {
  initial: number;
  min: number;
  max: number;
}

export default function useCounter(
  props: IUseCounterProps = {
    initial: 1,
    min: 1,
    max: 100,
  }
) {
  const [state, setState] = useState<IUseCounterSte>({ count: props.initial });


  const increment = () => {
    console.log('increment');
    setState({
      count: (state.count >= props.max) ? props.min : state.count + 1,
    });
  }

  const decrement = () => {
    setState({
      count: (state.count <= props.min) ? props.max : state.count - 1,
    });
  }

  const reset = () => {
    setState({
      count: props.initial,
    });
  }

  return {
    count: state.count,
    increment,
    decrement,
    reset,
  };
}
