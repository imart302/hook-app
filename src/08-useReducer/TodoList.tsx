import React from 'react';
import { ITodoItemProps, TodoItem } from './TodoItem';

export interface ITodoListProps {
  todos: ITodoItemProps[];
  onDeleteTodo?: (id: number) => void;
  onToggleTodo?: (id: number) => void;
}

export const TodoList = (props: ITodoListProps) => {
  const handleDeleteTodo = (id: number) => {
    props.onDeleteTodo?.(id);
  };

  const handleToggleTodo = (id: number) => {
    props.onToggleTodo?.(id);
  };

  return (
    <>
      <ul className="list-group">
        {props.todos.map((todo) => {
          return (
            <TodoItem
              key={todo.id}
              description={todo.description}
              id={todo.id}
              done={todo.done}
              onDeleteTodo={handleDeleteTodo}
              onToggleTodo={handleToggleTodo}
            ></TodoItem>
          );
        })}
      </ul>
    </>
  );
};
