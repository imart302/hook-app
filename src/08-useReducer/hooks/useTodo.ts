import React, { useReducer } from 'react'
import { ITodoActionReducer, todoReducer } from '../todoReducer';


interface INewTodo {
  description: string;
}

const initialState: never[] = [

];

const init = () => {
  return JSON.parse(localStorage.getItem('todos') ?? '[]');
}


export const useTodo = () => {

  const [state, dispatch] = useReducer(todoReducer, initialState, init);

  const todoCount = state.length;
  const pendingTodos = state.filter( (todo: { done: boolean; }) => todo.done==false).length;

  const onNewTodo = (todo: INewTodo) => {
    
    const newTodo = {
      id: new Date().getTime(),
      description: todo.description,
      done: false
    }

    const action: ITodoActionReducer = {
      type: '[TODO] Add Todo',
      payload: [newTodo],
    }

    dispatch(action);

  }

  const handleDeleteTodo = (id: number) => {
    const action: ITodoActionReducer = {
      type: '[TODO] Remove',
      payload: {
        id
      }
    }

    dispatch(action);
  }

  const handleToggleTodo = (id: number) => {
    console.log('toggle todo');
    const action: ITodoActionReducer = {
      type: '[TODO] Toggle',
      payload: {
        id
      }
    }

    dispatch(action);
  };
  return (
    {
      state,
      onNewTodo,
      handleDeleteTodo,
      handleToggleTodo,
      todoCount,
      pendingTodos,
    }
  )
}
