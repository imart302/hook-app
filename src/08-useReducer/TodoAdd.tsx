import React, { FormEvent, useState } from 'react';
import { useForm } from './hooks/useForm';
import { ITodoItemProps } from './TodoItem';

export interface INewTodo {
  description: string;
}

export interface ITodoAddSte {
  input: string;
}

export interface ITodoAddProps {
  onNewTodo?: (todo: INewTodo) => void;
}

export const TodoAdd = (props: ITodoAddProps) => {
  const { formFields, onInputChange } = useForm({
    fields: [{ name: 'description', value: '' }],
  });

  const onFormSubmit = (event: FormEvent) => {
    event.preventDefault();
    props.onNewTodo?.({
      description:
        formFields.find((field) => field.name === 'description')?.value ?? '',
    });
  };

  return (
    <>
      <form onSubmit={onFormSubmit}>
        <input
          type="text"
          placeholder="todo"
          className="form-control"
          value={
            formFields.find((field) => (field.name = 'description'))?.value
          }
          onChange={onInputChange}
          name="description"
        ></input>
        <button
          className="btn btn-primary mt-2"
          type="submit"
          onClick={onFormSubmit}
        >
          Agregar
        </button>
      </form>
    </>
  );
};
