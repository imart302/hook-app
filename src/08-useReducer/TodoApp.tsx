import React, { FormEvent, useEffect, useReducer } from 'react';
import { useTodo } from './hooks/useTodo';
import { TodoAdd } from './TodoAdd';
import { TodoList } from './TodoList';


export const TodoApp = () => {
  const { state, onNewTodo, handleDeleteTodo, handleToggleTodo, todoCount, pendingTodos } = useTodo();

  useEffect(() => {
    localStorage.setItem('todos', JSON.stringify(state));
    return () => {};
  }, [state]);

  return (
    <>
      <h1 data-testid="heding-count-id">
        TodoApp: <span data-testid="total-count-id">{todoCount}</span>, <small>pendientes: <span data-testid="pending-id">{pendingTodos}</span> </small>{' '}
      </h1>
      <hr />

      <div className="row">
        <div className="col-7">
          <TodoList
            data-testid="test-TodoList-id"
            todos={state}
            onDeleteTodo={handleDeleteTodo}
            onToggleTodo={handleToggleTodo}
          ></TodoList>
        </div>

        <div className="col-5">
          <h4>Agregar TODO</h4>
          <hr />
          <TodoAdd data-testid="test-TodoAdd-id" onNewTodo={onNewTodo}></TodoAdd>
        </div>
      </div>
    </>
  );
};
