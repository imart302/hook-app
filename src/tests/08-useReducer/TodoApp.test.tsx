import { vi, describe, expect } from 'vitest';
import { useTodo } from '../../08-useReducer/hooks/useTodo';
import { ITodoAddProps, TodoAdd } from '../../08-useReducer/TodoAdd';
import { TodoList } from '../../08-useReducer/TodoList';
import { render, screen } from '@testing-library/react';
import { TodoApp } from '../../08-useReducer/TodoApp';

const todos = [
  {
    id: 1,
    description: 'Desc1',
    done: false,
  },
  {
    id: 2,
    description: 'Desc2',
    done: false,
  }
]

vi.mock('../../08-useReducer/TodoAdd', () => {
  return {
    TodoAdd: vi.fn(() => {
      return (<div>TodoAddMock</div>);
    }),
  }
});

vi.mock('../../08-useReducer/TodoList', () => {
  return {
    TodoList: vi.fn(() => {
      return (<div>TodoListMock</div>);
    }),
  }
});

vi.mock('../../08-useReducer/hooks/useTodo', () => {
  return {
    useTodo: vi.fn(() => {

    }),
  }
});

describe('Test on Todo App', () => {

  const useTodoMocked = {
    state: todos,
    onNewTodo: vi.fn(),
    handleDeleteTodo: vi.fn(),
    handleToggleTodo: vi.fn(),
    todoCount: todos.length,
    pendingTodos: todos.length,
  }

  const TodoAddMockProps : ITodoAddProps = {
    onNewTodo: vi.fn(),
  }
  const TodoAddMock = vi.fn((TodoAddMockProps) => {
    return (<></>)
  })

  beforeAll(() => {
    vi.mocked(useTodo).mockReturnValue(useTodoMocked);
  });

  beforeEach(() => {

  });

  test('Should match snapshot', () => {
   
    const { container } = render(<TodoApp></TodoApp>);
    screen.debug();

    const countEl = screen.getByTestId('heding-count-id');

    // console.log(countEl.innerHTML);

    expect(container).toMatchSnapshot();
  });

  test('Should have 2  task and 2 pending', () => {

  });

});