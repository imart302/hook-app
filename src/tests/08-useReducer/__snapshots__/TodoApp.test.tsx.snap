// Vitest Snapshot v1

exports[`Test on Todo App > Should match snapshot 1`] = `
<div>
  <h1
    data-testid="heding-count-id"
  >
    TodoApp: 
    <span
      data-testid="total-count-id"
    >
      2
    </span>
    , 
    <small>
      pendientes: 
      <span
        data-testid="pending-id"
      >
        2
      </span>
       
    </small>
     
  </h1>
  <hr />
  <div
    class="row"
  >
    <div
      class="col-7"
    >
      <div>
        TodoListMock
      </div>
    </div>
    <div
      class="col-5"
    >
      <h4>
        Agregar TODO
      </h4>
      <hr />
      <div>
        TodoAddMock
      </div>
    </div>
  </div>
</div>
`;

exports[`Test on Todo App > Should render with components 1`] = `
<div>
  <h1
    data-testid="heding-count-id"
  >
    TodoApp: 
    <span
      data-testid="total-count-id"
    >
      2
    </span>
    , 
    <small>
      pendientes: 
      <span
        data-testid="pending-id"
      >
        2
      </span>
       
    </small>
     
  </h1>
  <hr />
  <div
    class="row"
  >
    <div
      class="col-7"
    >
      <div>
        TodoListMock
      </div>
    </div>
    <div
      class="col-5"
    >
      <h4>
        Agregar TODO
      </h4>
      <hr />
      <div>
        TodoAddMock
      </div>
    </div>
  </div>
</div>
`;
